parser.py

```diff
@@ -45,11 +45,22 @@ def format_wires(wire_id_list):
 
 class ArithmeticOp:
 
+    def __new__(cls, func, *exps):
+        if len(exps) == 1:
+            return super().__new__(UnaryOp)
+
+        if len(exps) == 2:
+            return super().__new__(BinaryOp)
+
+        return super().__new__(cls)
+
     def __init__(self, func, *exps):
+        self.func = func
+        self.args = exps
         self.tuple = (func, *exps)
 
     def __repr__(self):
-        return self.__str__()
```

Comment:
This was leading to a recursion error for me in the terminal. By default, if only `__repr__` is defined, than `__str__` will also work, since the default string method is
```python
def __str__(self):
    return self.__repr__()
```

So I've for now simply defined `__repr__` everywhere, and removed `__str__`